pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,805 @@
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
+		klog.ErrorS(err, "Couldn't create the directory for storing captured packets", "directory", packetDirectory)
+		return
+	}
+
+	// check the captures that are waiting in line.
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		if !apierrors.IsNotFound(err) {
+			c.queue.AddRateLimited(key)
+			klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	return true
+}
+
+func (c *Controller) syncPacketCapture(pcName string) error {
+	cleanupStatus := func() {
+		c.lock.Lock()
+		delete(c.captures, pcName)
+		c.lock.Unlock()
+	}
+
+	pc, err := c.packetCaptureLister.Get(pcName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.cleanupPacketCapture(pcName)
+			cleanupStatus()
+		}
+		return err
+	}
+
+	if err := c.validatePacketCapture(&pc.Spec); err != nil {
+		klog.ErrorS(err, "Invalid PacketCapture", "name", pc.Name)
+		if err := c.updatePacketCaptureStatus(pc.Name, 0, "", err); err != nil {
+			klog.ErrorS(err, "Failed to update PacketCapture status", "name", pc.Name)
+		}
+		cleanupStatus()
+		return nil
+	}
+
+	// Capture will not occur on this Node if a corresponding Pod interface is not found.
+	device := c.getTargetCaptureDevice(pc)
+	if device == nil {
+		return nil
+	}
+
+	status, err := func() (packetCapturePhase, error) {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		status := c.captures[pcName]
+		klog.InfoS("Syncing PacketCapture", "packagecapture", pcName, "status", status)
+		if status == packetCapturePhaseUnknown || status == packetCapturePhasePending {
+			klog.InfoS("Pending PackageCapture", "name", pcName)
+			newStatus, err := func() (packetCapturePhase, error) {
+				if c.numRunningCaptures < maxConcurrentCaptures {
+					c.numRunningCaptures += 1
+					return packetCapturePhaseRunning, c.startPacketCapture(pcName)
```

Comment:
It should only +1 after packet capture starts successfully, otherwise it would never -1.